src/queryHistory/queryHistoryProvider.ts

```diff
@@ -0,0 +1,206 @@
+    readonly onDidChangeTreeData: vscode.Event<any | undefined> = this._onDidChangeTreeData.event;
+
+    private _queryHistoryNodes: vscode.TreeItem[] = [new EmptyHistoryNode()]
+    private _queryHistoryLimit: number;
+    private _queryHistoryEnabled: boolean;
+    private _queryHistoryUI: QueryHistoryUI
+
+    constructor(
+        private _connectionManager: ConnectionManager,
+        private _outputContentProvider: SqlOutputContentProvider,
+        private _vscodeWrapper: VscodeWrapper,
+        private _untitledSqlDocumentService: UntitledSqlDocumentService,
+        private _statusView: StatusView,
+        private _prompter: IPrompter
+    ) {
+        const config = this._vscodeWrapper.getConfiguration(Constants.extensionConfigSectionName);
+        this._queryHistoryLimit = config.get(Constants.configQueryHistoryLimit);
+        this._queryHistoryEnabled = config.get(Constants.configEnableQueryHistoryCapture);
+        this._vscodeWrapper.setContext(Constants.isQueryHistoryEnabled, this._queryHistoryEnabled);
+        this._queryHistoryUI = new QueryHistoryUI(this._prompter, this._vscodeWrapper);
+    }
+
+    clearAll(): void {
+        this._queryHistoryNodes = [new EmptyHistoryNode()];
+        this._onDidChangeTreeData.fire();
+    }
+
+    refresh(ownerUri: string, timeStamp: Date, hasError): void {
+        const timeStampString = timeStamp.toLocaleString();
+        const historyNodeLabel = this.createHistoryNodeLabel(ownerUri, timeStampString);
+        const tooltip = this.createHistoryNodeTooltip(ownerUri, timeStampString);
+        const queryString = this.getQueryString(ownerUri);
+        const connectionLabel = this.getConnectionLabel(ownerUri);
+        const node = new QueryHistoryNode(historyNodeLabel, tooltip, queryString,
+            ownerUri, timeStamp, connectionLabel, !hasError);
+        if (this._queryHistoryNodes.length === 1) {
+            if (this._queryHistoryNodes[0] instanceof EmptyHistoryNode) {
+                this._queryHistoryNodes = [];
+            }
+        }
+        this._queryHistoryNodes.push(node);
+        // Push out the first listing if it crosses limit to maintain
+        // an LRU order
+        if (this._queryHistoryNodes.length > this._queryHistoryLimit) {
+            this._queryHistoryNodes.shift();
+        }
+        // return the query history sorted by timestamp
+        this._queryHistoryNodes.sort((a, b) => {
+            return (b as QueryHistoryNode).timeStamp.getTime()-
+                (a as QueryHistoryNode).timeStamp.getTime();
+        });
+        this._onDidChangeTreeData.fire();
+    }
+
+    getTreeItem(node: QueryHistoryNode): any {
+        return node;
+    }
+
+    async getChildren(element?: any): Promise<vscode.TreeItem[]> {
+        if (this._queryHistoryNodes.length == 0) {
+            this._queryHistoryNodes.push(new EmptyHistoryNode());
```

Comment:
Instead of making a new node each time why not just make one in the constructor and then if queryHistoryNodes is empty return an array that contains just that node instead? That saves you having to do the extra logic to clear it when adding a new node too.